### PR TITLE
Rebase branch with Outlook learning fixes

### PR DESCRIPTION
**Pull request by elie222**
# User description
This rebases the work onto latest main and carries the full branch diff onto the target branch. It adds Outlook category-reversal learning for MOVE_FOLDER actions, keeps webhook rate-limit wrapping, and limits learning to applied executions with added tests. It also includes automation-job execution/queue/messaging/stale handling updates and related assistant settings/tooling adjustments now present in the diff. A small Prisma filter typing fix is included by removing readonly const assertions.

---

# Generated description

Below is a concise technical summary of the changes proposed in this PR:
Refactor automation job execution by centralizing the work in <code>executeAutomationJobRun</code>, extending queue retry/backoff, multi-provider messaging, cron dispatch, and proactive-update settings/tools to honor broader channel requirements and stale-run handling. Extend Outlook category-reversal learning so applied label/folder undos feed <code>learnFromOutlookCategoryReversal</code>, keeping webhook rate limits, new edge-case guards, and richer tests that drive learned exclusions.
<table><tr><th>Topic</th><th>Details</th><tr><td><a href=https://baz.co/changes/elie222/inbox-zero/1793?tool=ast&topic=Outlook+learning>Outlook learning</a>
        </td><td>Broaden Outlook webhook learning by renaming to category reversal, filtering only applied executions, recognizing both label and MOVE_FOLDER undos (including folder-name-only), and preserving rate-limit instrumentation plus process-history wiring and extensive edge-case tests.<details><summary>Modified files (4)</summary><ul><li>apps/web/app/api/outlook/webhook/learn-label-removal.test.ts</li>
<li>apps/web/app/api/outlook/webhook/learn-label-removal.ts</li>
<li>apps/web/app/api/outlook/webhook/process-history.test.ts</li>
<li>apps/web/app/api/outlook/webhook/process-history.ts</li></ul></details><details><summary>Latest Contributors(1)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>elie222</td><td>webhook-reduce-blockin...</td><td>March 02, 2026</td></tr></table></details></td></tr>
<tr><td><a href=https://baz.co/changes/elie222/inbox-zero/1793?tool=ast&topic=Automation+job+flow>Automation job flow</a>
        </td><td>Revamp the automation job flow by routing queue and HTTP triggers through <code>executeAutomationJobRun</code>, handling stale runs, multi-provider messaging (<code>sendAutomationMessage</code>/<code>messaging-channel</code>), cron dispatch readiness checks, and proactive updates tooling/validation so scheduling honors any supported channel with proper labels, destinations, and tests (<code>messaging.test</code>, <code>stale.test</code>).<details><summary>Modified files (14)</summary><ul><li>apps/web/app/(app)/[emailAccountId]/assistant/settings/ProactiveUpdatesSetting.tsx</li>
<li>apps/web/app/api/automation-jobs/execute/queue/route.ts</li>
<li>apps/web/app/api/automation-jobs/execute/route.ts</li>
<li>apps/web/app/api/cron/automation-jobs/route.ts</li>
<li>apps/web/utils/actions/automation-jobs.ts</li>
<li>apps/web/utils/actions/automation-jobs.validation.ts</li>
<li>apps/web/utils/ai/assistant/chat-settings-tools.test.ts</li>
<li>apps/web/utils/ai/assistant/chat-settings-tools.ts</li>
<li>apps/web/utils/automation-jobs/execute.ts</li>
<li>apps/web/utils/automation-jobs/messaging-channel.ts</li>
<li>apps/web/utils/automation-jobs/messaging.test.ts</li>
<li>apps/web/utils/automation-jobs/messaging.ts</li>
<li>apps/web/utils/automation-jobs/stale.test.ts</li>
<li>apps/web/utils/automation-jobs/stale.ts</li></ul></details><details><summary>Latest Contributors(1)</summary><table><tr><th>User</th><th>Commit</th><th>Date</th></tr><tr><td>elie222</td><td>Fix-scheduled-check-in...</td><td>March 04, 2026</td></tr></table></details></td></tr></table>
This pull request is reviewed by Baz. Review like a pro on <a href=https://baz.co/changes/elie222/inbox-zero/1793?tool=ast>(Baz)</a>.

### CODE REVIEW

File: apps/web/utils/automation-jobs/execute.ts [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+import { getAutomationJobMessage } from "@/utils/automation-jobs/message";
+import type { Logger } from "@/utils/logger";
+import { isActivePremium } from "@/utils/premium";
+import prisma from "@/utils/prisma";
+import { getUserPremium } from "@/utils/user/get";
+import { sendAutomationMessage } from "@/utils/automation-jobs/messaging";
+
+export const executeAutomationJobBody = z.object({
+  automationJobRunId: z.string().min(1, "Automation job run ID is required"),
+});
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/automation-jobs/execute.ts]: Implemented. I moved `executeAutomationJobBody` to `apps/web/utils/actions/automation-jobs.validation.ts` and updated both execute routes to import the schema from the validation module.

File: apps/web/app/api/automation-jobs/execute/queue/route.ts [hunk 2]
```diff
@@ -1,19 +1,59 @@
...
+export const POST = handleCallback<z.infer<typeof executeAutomationJobBody>>(
+  async (message, metadata) => {
+    const parseResult = executeAutomationJobBody.safeParse(message);
+    if (!parseResult.success) {
+      logger.error("Invalid automation jobs queue payload", {
+        errors: parseResult.error.errors,
+        queueMessageId: metadata.messageId,
+      });
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/app/api/automation-jobs/execute/queue/route.ts]: Implemented. The queue route now creates its scoped logger inside the callback (per request/message context), and only derives child context from that local logger. There is no longer a module-global logger being enriched with `.with()`.

File: apps/web/utils/ai/assistant/chat-settings-tools.ts [hunk 3]
```diff
@@ -1132,16 +1150,23 @@ function buildScheduledCheckInsSnapshot(
...
   teamName,
 }: {
+  provider: MessagingProvider;
   channelName: string | null;
   teamName: string | null;
 }) {
   if (channelName && teamName) return `#${channelName} (${teamName})`;
   if (channelName) return `#${channelName}`;
-  return teamName || "Slack destination";
+  if (teamName) return teamName;
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/ai/assistant/chat-settings-tools.ts]: Implemented. I extracted a shared formatter in `apps/web/utils/automation-jobs/messaging-channel.ts` (`formatAutomationMessagingChannelLabel`) and switched both callers to use it.

File: apps/web/utils/automation-jobs/messaging.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  text,
+  logger,
+}: {
+  providerUserId: string | null;
+  text: string;
+  logger: Logger;
+}) {
+  if (!providerUserId) {
+    throw new AutomationJobConfigurationError(
+      "Teams channel is missing provider user ID",
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/automation-jobs/messaging.ts]: Implemented. I deduplicated the Teams/Telegram send flows by extracting `sendAutomationMessageToDirectAdapter` and made both provider-specific functions use it.

File: apps/web/app/api/cron/automation-jobs/route.ts [hunk 5]
```diff
@@ -84,6 +93,14 @@ async function enqueueDueAutomationJobs(logger: Logger) {
     let runId: string | null = null;
 
     try {
+      if (!isAutomationMessagingChannelReady(job.messagingChannel)) {
+        jobLogger.info(
+          "Skipped automation job because messaging channel is not ready",
+        );
+        skipped += 1;
+        continue;
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/app/api/cron/automation-jobs/route.ts:109]: Implemented. For unready messaging channels, we now advance `nextRunAt` via `postponeUnreadyAutomationJob` before continuing. This prevents perpetual reselection/skip loops for the same due record.

File: apps/web/app/api/cron/automation-jobs/route.ts [hunk 6]
```diff
@@ -84,6 +93,14 @@ async function enqueueDueAutomationJobs(logger: Logger) {
...
+      }
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/app/api/cron/automation-jobs/route.ts:110]: Implemented as part of the same change: unready jobs now have `nextRunAt` advanced before they are skipped, preventing repeated processing of the same due job.

File: apps/web/utils/automation-jobs/execute.ts [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+      await markAutomationJobRunSkipped({
+        automationJobRunId,
+        error: "Messaging channel is disconnected",
+      });
+
+      return new Response("Messaging channel disconnected", { status: 200 });
+    }
+
+    const provider =
+      run.automationJob.messagingChannel.emailAccount.account.provider;
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/automation-jobs/execute.ts:163]: I reviewed this one and did not apply a code change. In the current schema, `EmailAccount.account` is a required relation (not nullable), so `account` itself should not be null at runtime. The existing `!provider` guard remains in place for invalid provider values.

File: apps/web/app/(app)/[emailAccountId]/assistant/settings/ProactiveUpdatesSetting.tsx [hunk 8]
```diff
@@ -372,15 +370,19 @@ export function ProactiveUpdatesSetting() {
...
   teamName: string | null;
 }) {
   if (channel.channelName) return `#${channel.channelName}`;
   if (channel.channelId) return `Channel ${channel.channelId}`;
   if (channel.teamName) return channel.teamName;
-  return "Slack workspace";
+
+  if (channel.provider === "TEAMS") return "Teams destination";
+  if (channel.provider === "TELEGRAM") return "Telegram destination";
+  return "Slack destination";
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/app/(app)/[emailAccountId]/assistant/settings/ProactiveUpdatesSetting.tsx]: Implemented. The fallback label now comes from the shared formatter and returns `Slack workspace` for Slack channels without name metadata.

File: apps/web/utils/ai/assistant/chat-settings-tools.ts [hunk 9]
```diff
@@ -1132,16 +1150,23 @@ function buildScheduledCheckInsSnapshot(
...
+
+  if (provider === MessagingProvider.TEAMS) return "Teams destination";
+  if (provider === MessagingProvider.TELEGRAM) return "Telegram destination";
+
+  return "Slack destination";
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/ai/assistant/chat-settings-tools.ts]: Implemented. The chat settings path now uses the shared formatter, and Slack fallback is now `Slack workspace` as requested.

File: apps/web/app/api/automation-jobs/execute/queue/route.ts [hunk 10]
```diff
@@ -1,19 +1,59 @@
...
-  visibilityTimeoutSeconds: 290,
-  getLoggerContext: (payload) => ({
-    automationJobRunId: payload.automationJobRunId,
-  }),
-});
+    const parseResult = executeAutomationJobBody.safeParse(message);
+    if (!parseResult.success) {
+      logger.error("Invalid automation jobs queue payload", {
+        errors: parseResult.error.errors,
+      });
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/app/api/automation-jobs/execute/queue/route.ts:24]: I reviewed this and I’m intentionally not extracting a shared factory right now. This codebase guideline prefers avoiding premature abstraction, and the two queue handlers have different execution contracts (runner behavior and error semantics), so a shared generic wrapper would add indirection with limited payoff at this stage. If a third custom queue runner appears or these two converge further, I agree we should extract then.

File: apps/web/utils/automation-jobs/messaging-channel.ts [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+  if (
+    options?.includeTeamNameWithChannel &&
+    channel.channelName &&
+    channel.teamName
+  ) {
+    return `#${channel.channelName} (${channel.teamName})`;
+  }
+
+  if (channel.channelName) return `#${channel.channelName}`;
+  if (channel.channelId) return `Channel ${channel.channelId}`;
```
[the remark this answers is not in the document]
Comment (elie222) [reply apps/web/utils/automation-jobs/messaging-channel.ts]: Implemented. I updated `formatAutomationMessagingChannelLabel` so `channelId` is only used for non-Slack providers; Slack now falls back to `Slack workspace` when name metadata is missing. I also added a test case for Slack + channelId fallback in `utils/automation-jobs/messaging.test.ts` (commit `b6e9424`).
